File: internal/extension/captcha/recaptcha.go
package captcha

import (
	"bytes"
	"encoding/json"
	"github.com/elabosak233/cloudsdale/internal/extension/config"
	"io"
	"net/http"
)

type GoogleRecaptcha struct {
	URL       string
	SiteKey   string
	SecretKey string
	Threshold float64
}

func NewGoogleRecaptcha() ICaptcha {
	return &GoogleRecaptcha{
		URL:       config.AppCfg().Captcha.ReCaptcha.URL,
		SiteKey:   config.AppCfg().Captcha.ReCaptcha.SiteKey,
		SecretKey: config.AppCfg().Captcha.ReCaptcha.SecretKey,
		Threshold: config.AppCfg().Captcha.ReCaptcha.Threshold,
	}
}

func (g *GoogleRecaptcha) Verify(token string, clientIP string) (success bool, err error) {
	type RecaptchaRequest struct {
		Secret   string `json:"secret"`
		Response string `json:"response"`
		RemoteIP string `json:"remoteip"`
	}
	requestBody, err := json.Marshal(
		RecaptchaRequest{
			Secret:   g.SecretKey,
			Response: token,
			RemoteIP: clientIP,
		},
	)
	result, err := http.Post(g.URL, "application/json", bytes.NewBuffer(requestBody))
	defer func() {
		_ = result.Body.Close()
	}()
	body, err := io.ReadAll(result.Body)
	var response map[string]interface{}
	err = json.Unmarshal(body, &response)
	success, ok := response["success"].(bool)
	score, ok := response["score"].(float64)
	if ok && success && score >= g.Threshold {
		return true, err
	}
	return false, err
}

File: src/captcha/recaptcha.rs
use reqwest::Client;
use serde::Serialize;

use super::traits::ICaptcha;

pub struct Recaptcha {
    secret_key: String,
    url: String,
    threshold: f64,
}

#[derive(Serialize)]
struct RecaptchaRequest {
    #[serde(rename = "secret")]
    secret_key: String,
    #[serde(rename = "response")]
    response: String,
    #[serde(rename = "remoteip")]
    remote_ip: String,
}

impl ICaptcha for Recaptcha {
    fn new() -> Self {
        return Recaptcha {
            url: crate::config::get_app_config()
                .captcha
                .recaptcha
                .url
                .clone(),
            secret_key: crate::config::get_app_config()
                .captcha
                .recaptcha
                .secret_key
                .clone(),
            threshold: crate::config::get_app_config().captcha.recaptcha.threshold,
        };
    }

    async fn verify(&self, token: String, client_ip: String) -> bool {
        let request_body = RecaptchaRequest {
            secret_key: self.secret_key.clone(),
            response: token,
            remote_ip: client_ip,
        };

        let client = Client::new();
        let resp = client
            .post(&self.url)
            .json(&request_body)
            .send()
            .await
            .unwrap();

        let response: serde_json::Value = resp.json().await.unwrap();

        match response.get("success") {
            Some(result) => {
                if let serde_json::Value::Bool(success) = result {
                    match response.get("score") {
                        Some(score) => {
                            if let serde_json::Value::Number(score) = score {
                                let score = score.as_f64().unwrap_or(0.0);
                                if *success && score >= self.threshold {
                                    return true;
                                } else {
                                    return false;
                                }
                            }
                            return false;
                        }
                        None => return false,
                    }
                }
                return false;
            }
            None => return false,
        }
    }
}
